rapids-cmake/cpm/detail/get_proprietary_binary.cmake

```diff
@@ -0,0 +1,78 @@
+#=============================================================================
+# Copyright (c) 2022, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+include_guard(GLOBAL)
+
+#[=======================================================================[.rst:
+get_proprietary_binary
+-------------------
+
+.. versionadded:: v22.06.00
+
+Download the associated proprietary binary for the given project based on
+the current CPU target architecture ( x86_64, aarch64, etc )
+
+ .. note::
+  if override => the proprietary entry only in the override will be evaluated
+  if no override => the proprietary entry only in the default will be evaluated
+
+
+#]=======================================================================]
+function(rapids_cpm_get_proprietary_binary package_name version)
+
+  get_property(override_json_data GLOBAL PROPERTY rapids_cpm_${package_name}_override_json)
+  # need to search the `proprietary_binary` dictonary for a key with the same name as
+  # lower_case(`CMAKE_SYSTEM_PROCESSOR-CMAKE_SYSTEM_NAME`).
+  set(key "${CMAKE_SYSTEM_PROCESSOR}-${CMAKE_SYSTEM_NAME}")
+  string(TOLOWER ${key} key)
+
+  if(override_json_data)
+    string(JSON value ERROR_VARIABLE have_error GET "${override_json_data}" "proprietary_binary"
+           "${key}")
+  else()
+    get_property(json_data GLOBAL PROPERTY rapids_cpm_${package_name}_json)
+    string(JSON value ERROR_VARIABLE have_error GET "${json_data}" "proprietary_binary" "${key}")
+  endif()
+
+  if(have_error)
+    message(VERBOSE
+            "${package_name} requested usage of a proprietary_binary but none exist for ${CMAKE_SYSTEM_PROCESSOR}"
+    )
+    return()
+  endif()
+
+  # proprietary_binary now is the url that we need to download
+  set(proprietary_binary ${value})
+
+  if(NOT DEFINED rapids-cmake-version)
+    include("${rapids-cmake-dir}/rapids-version.cmake")
+  endif()
+
+  # Evaluate any magic placeholders in the proprietary_binary value including the
+  # `rapids-cmake-version` value
+  cmake_language(EVAL CODE "set(proprietary_binary ${proprietary_binary})")
```

Comment:
Would it be possible to instead evaluate a subset of properties across all of versions.json inside `rapids_cpm_load_preset_versions`? Or will that fail because the variables that aren't defined will get replaced with an empty string, e.g. `${version}` will end up empty because only `rapids-cmake-version` can be defined at the point that `rapids_cpm_load_preset_versions` is called? Trying to find ways to remove confusing non-local boilerplate code.

Reply:
We defer evaluation of placeholders as long as possible. That allows overrides to update only something like the `version` value and have that reflected in evaluations of other fields.

If we went to an eager model we would need to re-evaluate all previous entries excluding the keys that are part of the override set, which seemed harder to implement.